### PR TITLE
Feature/passer and receiver roles

### LINKED ISSUE #899: PassRole

### LINKED ISSUE #902: PassReceiverRole

### LINKED ISSUE #906: Better check for when we need to reassign roles
We currently reassign roles whenever the amount of STPInfos is different from the amount of robots currently belonging to us. 

This means that if more STPInfos are assigned it will keep reassigning roles. 

Suggestion: have the play only reassign roles at the start of the play, and then whenever we lose a robot due to yellow card (which we will know because it will go through stopstate so we can make that our invariant) we recalculate the play, and when we lose a robot due to vision we don't recalculate all roles as we expect it to come back soon.

### PR DESCRIPTION
**Pull request by RobotJente**
This branch includes a Pass Play and Passer and Attacker Roles

### CODE REVIEW

File: include/roboteam_ai/stp/Role.hpp [hunk 1]
```diff
@@ -59,6 +59,11 @@ class Role {
      * @return true if the active tactic cannot execute (it's prerequisites are no longer met)
      */
     virtual bool shouldRoleReset(const StpInfo &info) noexcept = 0;
+
+    /**
+     * Current status of tactic state machine
+     */
+     Status currentTacticStatus;
```
Comment (timovdk) [inline include/roboteam_ai/stp/Role.hpp]: Why do you need to store the current tactic status as a member?

File: include/roboteam_ai/stp/new_plays/Pass.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+#define RTT_PASS_H
+
+#include <stp/Play.hpp>
+
+namespace rtt::ai::stp {
+
+    class Pass : public Play {
+    public:
+        /**
+         * Constructor that initializes roles with test roles
```
Comment (timovdk) [inline include/roboteam_ai/stp/new_plays/Pass.h]:
```suggestion
         * Constructor that initializes roles with roles that are necessary for this play
```

File: include/roboteam_ai/stp/new_plays/Pass.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        uint8_t score(world_new::World* world) noexcept override;
+
+        /**
+         * Assigns robots to roles of this play
+         */
+        void assignRoles() noexcept;
+
+        void calculateInfoForPlay() noexcept override;
+
+        std::vector<Vector2> calculateDefensivePositions(int numberOfDefenders, world_new::World *world,
```
Comment (timovdk) [inline include/roboteam_ai/stp/new_plays/Pass.h]: These probably need comments

File: include/roboteam_ai/stp/Role.hpp (hunk 1 above)
Comment (RobotJente) [reply include/roboteam_ai/stp/Role.hpp]: I don't think I do haha, I forgot to remove this

File: src/stp/Role.cpp [hunk 4]
```diff
@@ -13,21 +13,25 @@ Status Role::update(StpInfo const& info) noexcept {
...
     }
 
+    // Update tactic info
+    auto tacticInfo = calculateInfoForTactic(info);
+
     // Reset the role
-    if (robotTactics.current_num() != 0 && shouldRoleReset(info)) {
+    if ((robotTactics.current_num() != 0) && shouldRoleReset(tacticInfo)) {
         RTT_INFO("State Machine reset for current role for ID = ", info.getRobot()->get()->getId())
         // TODO: messy reset, do it in the state machine
```
Comment (timovdk) [inline src/stp/Role.cpp]:
```suggestion
```

Comment (timovdk) [reply src/stp/Role.cpp]: The reset is not messy anymore, yay!

File: src/stp/Role.cpp [hunk 5]
```diff
@@ -13,21 +13,25 @@ Status Role::update(StpInfo const& info) noexcept {
...
+        robotTactics.get_current()->skills.reset();
     }
 
-    // Update tactic info
-    auto tacticInfo = calculateInfoForTactic(info);
-
     // Update the state machine of tactics with the TacticInfo from Play
-    return robotTactics.update(tacticInfo);
+    currentTacticStatus = robotTactics.update(tacticInfo);
+    return currentTacticStatus;
```
Comment (timovdk) [inline src/stp/Role.cpp]: Why assign and return? No need for this variable right?

File: include/roboteam_ai/stp/new_roles/PassReceiver.h [hunk 6]
```diff
@@ -0,0 +1,28 @@
+//
+// Created by jessevw on 17.03.20.
+//
+
+#ifndef RTT_PASSRECEIVER_H
+#define RTT_PASSRECEIVER_H
+
+#include <stp/Role.hpp>
+
+namespace rtt::ai::stp {
```
Comment (Ratoone) [inline include/roboteam_ai/stp/new_roles/PassReceiver.h]:
```suggestion
namespace rtt::ai::stp::role {
```

File: include/roboteam_ai/stp/new_plays/Pass.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
+//
+// Created by jessevw on 17.03.20.
+//
+
+#ifndef RTT_PASS_H
+#define RTT_PASS_H
+
+#include <stp/Play.hpp>
+
+namespace rtt::ai::stp {
```
Comment (Ratoone) [inline include/roboteam_ai/stp/new_plays/Pass.h]:
```suggestion
namespace rtt::ai::stp::play {
```

File: src/stp/Role.cpp [hunk 8]
```diff
@@ -13,21 +13,25 @@ Status Role::update(StpInfo const& info) noexcept {
...
+
     // Reset the role
-    if (robotTactics.current_num() != 0 && shouldRoleReset(info)) {
+    if ((robotTactics.current_num() != 0) && shouldRoleReset(tacticInfo)) {
         RTT_INFO("State Machine reset for current role for ID = ", info.getRobot()->get()->getId())
         // TODO: messy reset, do it in the state machine
-        robotTactics.skip_n(-robotTactics.current_num());
+        robotTactics.reset();
+        RTT_INFO("current state of SM: ", robotTactics.current_num(), roleName)
+        robotTactics.get_current()->skills.reset();
```
Comment (Ratoone) [inline src/stp/Role.cpp]: tactic.reset()

File: include/roboteam_ai/stp/Role.hpp [hunk 9]
```diff
@@ -59,6 +59,11 @@ class Role {
      * @return true if the active tactic cannot execute (it's prerequisites are no longer met)
      */
     virtual bool shouldRoleReset(const StpInfo &info) noexcept = 0;
+
+    /**
+     * Stores the current status of tactic to be able to decide whether to reset or not.
+     */
+    Status currentTacticStatus;
```
Comment (Ratoone) [inline include/roboteam_ai/stp/Role.hpp]: Should not be needed as you get the status after update

**Comment by timovdk** [comment]
Closes #899 #902 #906

File: src/stp/new_plays/Pass.cpp [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+void Pass::calculateInfoForRole() noexcept {
+    // Calculate most important positions to defend
+    // You know you have n defenders, because the play assigned it that way
+    auto enemyRobots = world->getWorld()->getThem();
+    auto defensivePositions = calculateDefensivePositions(2, world, enemyRobots);
+
+    // Use this new information to assign the roles using the dealer.
+    decltype(stpInfos)::iterator stpInfos_iter{};
```
Comment (Ratoone) [inline src/stp/new_plays/Pass.cpp]: Uhm, I guess this is an option, but can't we do it somehow else? I'm not a very big fan o huge switch statements, but maybe that works better for now

File: src/stp/new_roles/PassReceiver.cpp [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace rtt::ai::stp {
+
+    PassReceiver::PassReceiver(std::string name) : Role(std::move(name)) {
+        // create state machine and initializes the first state
+        robotTactics = collections::state_machine<Tactic, Status, StpInfo>{tactic::Receive()};
+        robotTactics.initialize();
+    }
+
+    StpInfo PassReceiver::calculateInfoForTactic(StpInfo const &info) noexcept {
```
Comment (Ratoone) [inline src/stp/new_roles/PassReceiver.cpp]: Roles don't really compute anything, so maybe we don't need this

File: src/stp/new_roles/PassReceiver.cpp [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+        return info;
+    }
+
+    bool PassReceiver::shouldRoleReset(const StpInfo &info) noexcept {
```
Comment (Ratoone) [inline src/stp/new_roles/PassReceiver.cpp]: Same comment as above. Check the architecture as well

File: src/stp/new_tactics/KickAtPos.cpp [hunk 13]
```diff
@@ -40,14 +40,16 @@ StpInfo KickAtPos::calculateInfoForSkill(StpInfo const &info) noexcept {
 
     // When rotating, we need to dribble to keep the ball, but when kicking we don't
     if (skills.current_num() == 0) {
```
Comment (Ratoone) [inline src/stp/new_tactics/KickAtPos.cpp]: Suggestion: if robot is not facing the shooting position, dribble. It makes more sense logic-wise

**Review by Ratoone — CHANGES_REQUESTED** [review]

File: include/roboteam_ai/stp/Role.hpp (hunk 9 above)
Comment (timovdk) [reply include/roboteam_ai/stp/Role.hpp]: True, I think I will fix this in this PR after all

File: src/stp/new_plays/Pass.cpp (hunk 10 above)
Comment (timovdk) [reply src/stp/new_plays/Pass.cpp]: Yeah we need to think about the best way to do this, there is probably a better way. I will add a TODO.

File: src/stp/new_roles/PassReceiver.cpp (hunk 11 above)
Comment (timovdk) [reply src/stp/new_roles/PassReceiver.cpp]: True, I will leave it for now but add a todo so we don't forget

File: src/stp/new_roles/PassReceiver.cpp (hunk 12 above)
Comment (timovdk) [reply src/stp/new_roles/PassReceiver.cpp]: added todo

File: src/stp/new_plays/Pass.cpp (hunk 10 above)
Comment (RobotJente) [reply src/stp/new_plays/Pass.cpp]: Do what? The dealer assignment?

File: src/stp/new_tactics/KickAtPos.cpp (hunk 13 above)
Comment (timovdk) [reply src/stp/new_tactics/KickAtPos.cpp]: Done

File: src/stp/new_plays/Pass.cpp (hunk 10 above)
Comment (timovdk) [reply src/stp/new_plays/Pass.cpp]: The assignment of role-specific data

File: src/stp/new_roles/PassReceiver.cpp (hunk 11 above)
Comment (timovdk) [reply src/stp/new_roles/PassReceiver.cpp]: Actually removed it since we didn't need it

File: src/stp/new_roles/PassReceiver.cpp (hunk 12 above)
Comment (timovdk) [reply src/stp/new_roles/PassReceiver.cpp]: Actually removed it since we didn't need it

File: include/roboteam_ai/stp/new_plays/TestPlay.h [hunk 14]
```diff
@@ -39,7 +39,7 @@ class TestPlay : public Play {
      */
     void assignRoles() noexcept;
 
-    void calculateInfoForPlay() noexcept override;
+    void calculateInfoForRole() noexcept override;
```
Comment (RobotJente) [inline include/roboteam_ai/stp/new_plays/TestPlay.h]: Should we not call this calculateInfoForPlay?

File: src/stp/new_plays/Pass.cpp [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    auto enemyRobots = world->getWorld()->getThem();
+    auto defensivePositions = calculateDefensivePositions(2, world, enemyRobots);
+
+    // TODO: is there really no better way to set data per role?
+    // Use this new information to assign the roles using the dealer.
+    decltype(stpInfos)::iterator stpInfos_iter{};
+
+    // Calculate receiver info
+    stpInfos_iter = stpInfos.find("pass_receiver");
+    if (stpInfos_iter != stpInfos.end()) stpInfos_iter->second.setPosition({RECEIVE_AT_POSITION, Vector2(-2, -2)});
```
Comment (RobotJente) [inline src/stp/new_plays/Pass.cpp]:
```suggestion
    if (stpInfos.find("pass_receiver") != stpInfos.end()) {
        stpInfos["pass_receiver"].setPosition({RECEIVE_AT_POSITION, Vector2(-2, -2)});
    }
```

**Review by RobotJente — COMMENTED** [review]
I approve

File: src/stp/new_plays/Pass.cpp (hunk 15 above)
Comment (timovdk) [reply src/stp/new_plays/Pass.cpp]: Done

**Review by timovdk — APPROVED** [review]

**Review by Ratoone — APPROVED** [review]

**Pull request merged**